src/log-store/src/kafka/log_store.rs

```diff
@@ -148,26 +154,49 @@ impl LogStore for KafkaLogStore {
                     actual: entry.provider().type_name(),
                 })?
                 .clone();
-            producers
-                .entry(entry.region_id())
-                .or_insert_with(|| RecordProducer::new(provider))
-                .push(entry);
+
+            let region_id = entry.region_id();
+
+            if let hash_map::Entry::Vacant(e) = region_producers.entry(region_id) {
+                let producer = self
+                    .client_manager
+                    .get_or_insert(&provider)
+                    .await?
+                    .producer()
+                    .clone();
+                e.insert(producer);
+            };
+            region_grouped_records
+                .entry(region_id)
+                .or_default()
+                .extend(convert_to_kafka_records(entry)?);
         }
 
-        // Produces entries for each region and gets the offset those entries written to.
-        // The returned offset is then converted into an entry id.
-        let last_entry_ids = futures::future::try_join_all(producers.into_iter().map(
-            |(region_id, producer)| async move {
-                let entry_id = producer
-                    .produce(&self.client_manager)
-                    .await
-                    .map(TryInto::try_into)??;
-                Ok((region_id, entry_id))
-            },
-        ))
-        .await?
-        .into_iter()
-        .collect::<HashMap<_, _>>();
+        let mut region_ids = Vec::with_capacity(region_grouped_records.keys().len());
```

Comment:
We can calculate the region ids at the beginning of the method, so that `region_grouped_records` and `region_producers` can be constructed using with_capacity.